### PR TITLE
Apply dominance rule in constant distributions

### LINKED ISSUE #36: Constant distribution -> include dominance & overload default distributions

### PR DESCRIPTION
**Pull request by vankesteren**
fixes #36

Here's a test script:

```py
from metasyn import MetaFrame, VarSpec
from metasyncontrib.disclosure import DisclosurePrivacy
import polars as pl

df = (
    pl.DataFrame({"id": range(30)})
    .with_columns(
        pl.lit(1.0).alias("TESTCONT"),
        pl.lit(8).alias("TESTDISC"),
        pl.lit("Secret string").alias("TESTSTR"),
        pl.lit("2021-02-03").str.to_date().alias("TESTDATE"),
        pl.lit("12:35:01").str.to_time().alias("TESTTIME"),
        pl.lit("2021-02-03T12:35:01").str.to_datetime().alias("TESTDATETIME"),
    )
)

mf = MetaFrame.fit_dataframe(df, privacy=DisclosurePrivacy(), var_specs=[VarSpec("id", unique=True)])

# original data
df

# synthetic data
mf.synthesize(30)
```

## Original data
```
┌─────┬──────────┬──────────┬───────────────┬────────────┬──────────┬─────────────────────┐
│ id  ┆ TESTCONT ┆ TESTDISC ┆ TESTSTR       ┆ TESTDATE   ┆ TESTTIME ┆ TESTDATETIME        │
│ --- ┆ ---      ┆ ---      ┆ ---           ┆ ---        ┆ ---      ┆ ---                 │
│ i64 ┆ f64      ┆ i32      ┆ str           ┆ date       ┆ time     ┆ datetime[μs]        │
╞═════╪══════════╪══════════╪═══════════════╪════════════╪══════════╪═════════════════════╡
│ 0   ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 1   ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 2   ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 3   ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 4   ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ …   ┆ …        ┆ …        ┆ …             ┆ …          ┆ …        ┆ …                   │
│ 25  ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 26  ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 27  ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 28  ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
│ 29  ┆ 1.0      ┆ 8        ┆ Secret string ┆ 2021-02-03 ┆ 12:35:01 ┆ 2021-02-03 12:35:01 │
└─────┴──────────┴──────────┴───────────────┴────────────┴──────────┴─────────────────────┘
```

## Synthetic
```
┌─────┬──────────┬──────────┬──────────┬────────────┬──────────┬─────────────────────┐
│ id  ┆ TESTCONT ┆ TESTDISC ┆ TESTSTR  ┆ TESTDATE   ┆ TESTTIME ┆ TESTDATETIME        │
│ --- ┆ ---      ┆ ---      ┆ ---      ┆ ---        ┆ ---      ┆ ---                 │
│ i64 ┆ f64      ┆ i64      ┆ str      ┆ date       ┆ time     ┆ datetime[μs]        │
╞═════╪══════════╪══════════╪══════════╪════════════╪══════════╪═════════════════════╡
│ 0   ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 1   ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 2   ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 3   ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 4   ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ …   ┆ …        ┆ …        ┆ …        ┆ …          ┆ …        ┆ …                   │
│ 25  ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 26  ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 27  ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 28  ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
│ 29  ┆ 99999.9  ┆ 99999    ┆ REDACTED ┆ 1970-01-01 ┆ 00:00:00 ┆ 1970-01-01 00:00:00 │
└─────┴──────────┴──────────┴──────────┴────────────┴──────────┴─────────────────────┘

### CODE REVIEW

File: metasyncontrib/disclosure/base.py [hunk 1]
```diff
@@ -30,6 +30,13 @@ def fit(cls, series, *args, n_avg: int = 11, **kwargs) -> BaseDistribution:
         """Fit constant distributions with disclosure control rules in place."""
         pl_series: pl.Series = cls._to_series(series)
 
+        # NB: dominance rule ensures that constant distribution is essentially never
+        # allowed under formal disclosure control. Always return default distribution.
+        apply_dominance_rule: bool = True
```
Comment (qubixes) [inline metasyncontrib/disclosure/base.py:35]: I think we can remove the old code that won't be reached now. We have git in case we would like to get back to the old way.

**Review by qubixes — CHANGES_REQUESTED** [review]
I also was working on this, but obviously doing basically the same as in this PR. One thing that I did change differently was the MultinoulliDistribution. The dominance distribution is already in there, but it would be nice to align it with the default_distributions from the constant distributions.

File: metasyncontrib/disclosure/base.py (hunk 1 above)
Comment (vankesteren) [reply metasyncontrib/disclosure/base.py:35]: yeah I was preparing for a potential "apply dominance" parameter in `DisclosurePrivacy()`

File: tests/test_constant.py [hunk 2]
```diff
@@ -19,19 +19,18 @@
...
     ],
 )
-def test_constant(dist_builtin, dist_disclosure, value):
+def test_constant(dist_builtin, dist_disclosure, value, disclosurevalue):  # noqa: D103
     dist = dist_builtin(value)
     data = [dist.draw() for _ in range(21)]
 
-    assert dist_disclosure.fit(data, partition_size=22)._param_dict().get("value") != value
-    assert dist_disclosure.fit(data, partition_size=11)._param_dict().get("value") == value
+    assert dist_disclosure.fit(data, partition_size=11)._param_dict().get("value") == disclosurevalue
```
Comment (qubixes) [inline tests/test_constant.py:36]: Even though not part of this PR, I think this should work as well.
```suggestion
    assert dist_disclosure.fit(data, partition_size=11).value == disclosurevalue
```

**Review by qubixes — APPROVED** [review]
Thanks for adding the test. Approved now.
